django/db/models/sql/compiler.py

```diff
@@ -802,6 +810,54 @@ def get_related_klass_infos(klass_info, related_klass_infos):
                 )
         return related_klass_infos
 
+    def get_select_for_update_of_arguments(self):
+        """
+        Return a quoted list of arguments to the SELECT FOR UPDATE OF
+        part of the query.
+        """
+        def _get_field_choices():
+            # Yield all allowed field paths in BFS order.
+            queue = collections.deque([(None, self.klass_info)])
+            while queue:
+                parent_path, klass_info = queue.popleft()
+                if parent_path is None:
+                    path = []
+                    yield 'self'
+                else:
+                    path = parent_path + [klass_info['field'].name]
+                    yield LOOKUP_SEP.join(path)
+                queue.extend(
+                    (path, klass_info)
+                    for klass_info in klass_info.get('related_klass_infos', [])
+                )
+
+        result = []
+
+        for name in self.query.select_for_update_of:
+            parts = [] if name == 'self' else name.split(LOOKUP_SEP)
+            klass_info = self.klass_info
+            for part in parts:
+                for related_klass_info in klass_info.get('related_klass_infos', []):
+                    if related_klass_info['field'].name == part:
+                        klass_info = related_klass_info
+                        break
+                else:
+                    raise FieldError(
```

Comment:
1. It would be very difficult to move this error to earlier, since I rely on the `klass_info` data structure.

2. If `'self'` is an actual valid name, then yes. The meaning of `'self'` in `select_for_update(of=(..., 'self', ...))` is always the "root" table.